src/glyph_item_iter.rs

```diff
@@ -0,0 +1,207 @@
+    pub struct GlyphItemIter(Boxed<pango_sys::PangoGlyphItemIter>);
+
+    match fn {
+        copy => |ptr| pango_sys::pango_glyph_item_iter_copy(mut_override(ptr)),
+        free => |ptr| pango_sys::pango_glyph_item_iter_free(ptr),
+        init => |_ptr| (),
+        clear => |_ptr| (),
+        get_type => || pango_sys::pango_glyph_item_iter_get_type(),
+    }
+}
+
+impl GlyphItemIter {
+    pub(crate) unsafe fn init_end(glyph_item: &GlyphItem, text: &str) -> Option<GlyphItemIter> {
+        let mut iter = GlyphItemIter::uninitialized();
+        let ret = from_glib(pango_sys::pango_glyph_item_iter_init_end(
+            iter.to_glib_none_mut().0,
+            mut_override(glyph_item.to_glib_none().0),
+            text.to_glib_none().0,
+        ));
+        if ret {
+            Some(iter)
+        } else {
+            None
+        }
+    }
+
+    pub(crate) unsafe fn init_start(glyph_item: &GlyphItem, text: &str) -> Option<GlyphItemIter> {
+        let mut iter = GlyphItemIter::uninitialized();
+        let ret = from_glib(pango_sys::pango_glyph_item_iter_init_start(
+            iter.to_glib_none_mut().0,
+            mut_override(glyph_item.to_glib_none().0),
+            text.to_glib_none().0,
+        ));
+        if ret {
+            Some(iter)
+        } else {
+            None
+        }
+    }
+
+    pub fn next_cluster(&mut self) -> bool {
+        unsafe {
+            from_glib(pango_sys::pango_glyph_item_iter_next_cluster(
+                self.to_glib_none_mut().0,
+            ))
+        }
+    }
+
+    pub fn prev_cluster(&mut self) -> bool {
+        unsafe {
+            from_glib(pango_sys::pango_glyph_item_iter_prev_cluster(
+                self.to_glib_none_mut().0,
+            ))
+        }
+    }
+
+    pub fn to_data(&self) -> GlyphItemIteratorData {
+        GlyphItemIteratorData {
+            start_glyph: self.0.start_glyph,
+            end_glyph: self.0.end_glyph,
+            start_index: self.0.start_index as usize,
+            end_index: self.0.end_index as usize,
+            start_char: self.0.start_char as usize,
+            end_char: self.0.end_char as usize,
+        }
+    }
+}
+
+pub struct GlyphItemIteratorData {
+    pub start_glyph: i32,
+    pub start_index: usize,
+    pub start_char: usize,
+
+    pub end_glyph: i32,
+    pub end_index: usize,
+    pub end_char: usize,
+}
+
+enum NormalIterator {}
+enum ReverseIterator {}
+
+struct GlyphItemIterator<'a, T> {
+    item: &'a GlyphItem,
+    text: &'a str,
+    iter: Option<GlyphItemIter>,
+    _mode: PhantomData<T>,
+}
+
+impl GlyphItem {
+    pub fn iter<'a>(
+        &'a self,
+        text: &'a str,
+    ) -> impl DoubleEndedIterator<Item = GlyphItemIteratorData> + 'a {
+        GlyphItemIterator::<NormalIterator> {
+            item: self,
+            text,
+            iter: None,
+            _mode: PhantomData,
+        }
+    }
+
+    pub fn reverse_iter<'a>(
+        &'a self,
+        text: &'a str,

```

Comment:
How do we ensure that the text corresponds to the glyph item? From reading the code it seems like we otherwise would get invalid memory accesses (if the text is smaller than the glyph)

Reply:
I don't know way to ensure :(

Reply:
We can get PangoItem.length and compare with text length and chars count, but it not give us 100%

Reply:
I don't know either, needs someone to dig into this API

Reply:
@EPashkin This one is the remaining issue? I guess this needs some talking to Pango developers to figure out what to do about this here :)

Reply:
IMHO yes

Reply:
Any news on this?

Reply:
@GuillaumeGomez Sorry, but I still have no idea how do this safe,
so better just leave it for now.

Reply:
:+1: